st2actions/st2actions/utils/param_utils.py

```diff
@@ -45,295 +43,158 @@ def pf(params, skips):
     return (pf(runner_parameters, {}), pf(action_parameters, runner_parameters))
 
 
-def _get_resolved_runner_params(runner_parameters, action_parameters,
-                                actionexec_runner_parameters):
-    # Runner parameters should use the defaults from the RunnerType object.
-    # The runner parameter defaults may be overridden by values provided in
-    # the Action and liveaction.
+def _cast_params(rendered, parameter_schemas):
+    '''
+    It's just here to make tests happy
+    '''
+    casted_params = {}
+    for k, v in six.iteritems(rendered):
+        casted_params[k] = _cast(v, parameter_schemas[k] or {})
+    return casted_params
 
-    # Create runner parameter by merging default values with dynamic values
-    resolved_params = {k: v['default'] if 'default' in v else None
-                       for k, v in six.iteritems(runner_parameters)}
 
-    # pick overrides from action_parameters & actionexec_runner_parameters
-    for param_name, param_value in six.iteritems(runner_parameters):
-        # No override if param is immutable
-        if param_value.get('immutable', False):
-            continue
+def _cast(v, parameter_schema):
+    if v is None or not parameter_schema:
+        return v
 
-        # Check if param exists in action_parameters and if it has a default value then
-        # pickup the override.
-        if param_name in action_parameters:
-            action_param = action_parameters[param_name]
-            if 'default' in action_param:
-                resolved_params[param_name] = action_param['default']
+    parameter_type = parameter_schema.get('type', None)
+    if not parameter_type:
+        return v
 
-            # No further override (from liveaction) if param is immutable
-            if action_param.get('immutable', False):
-                continue
+    cast = get_cast(cast_type=parameter_type)
+    if not cast:
+        return v
 
-        # Finally pick up override from actionexec_runner_parameters
-        if param_name in actionexec_runner_parameters:
-            resolved_params[param_name] = actionexec_runner_parameters[param_name]
+    return cast(v)
 
-    return resolved_params
 
+def process(G, name, value):
```

Comment:
Sure, I'll add them. This was a repeatable pattern across the whole implementation so I've made it a method the last moment so i don't have to fix it in multiple places. What it does is that it creates a graph node corresponding to the parameter. It will either be a value node, a root of dependency subtree or a template node - a branch to another value node.